File: examples/nodejs/quickstart/quickstart.js
// voxrt-wake-word — Node.js quickstart over a 16 kHz WAV file.
//
// Loads a .vxrt, walks a WAV file in 32 ms chunks, prints every wake
// detection + final RTF. Pure stdlib (no npm deps beyond
// @voxrt/wake-word itself) — uses Buffer + manual WAV header parse.
//
// Usage:
//   node quickstart.js voxrt_wake_word.vxrt hey_assistant.wav
//
// WAV must be 16 kHz mono 16-bit. Convert with:
//   ffmpeg -i input.flac -ar 16000 -ac 1 -sample_fmt s16 output.wav

"use strict";

const fs = require("fs");
const path = require("path");
const { WakeWordEngine } = require("@voxrt/wake-word");

const CHUNK_FRAMES = 512;  // 32 ms at 16 kHz

function readWavMono16(filePath) {
  const buf = fs.readFileSync(filePath);
  if (buf.toString("ascii", 0, 4) !== "RIFF" ||
      buf.toString("ascii", 8, 12) !== "WAVE") {
    throw new Error(`${filePath}: not a RIFF/WAVE file`);
  }
  // Walk chunks to find `fmt ` and `data`.
  let pos = 12;
  let sampleRate = 0;
  let channels = 0;
  let bitsPerSample = 0;
  let dataOffset = 0;
  let dataSize = 0;
  while (pos < buf.length) {
    const id = buf.toString("ascii", pos, pos + 4);
    const size = buf.readUInt32LE(pos + 4);
    if (id === "fmt ") {
      channels = buf.readUInt16LE(pos + 10);
      sampleRate = buf.readUInt32LE(pos + 12);
      bitsPerSample = buf.readUInt16LE(pos + 22);
    } else if (id === "data") {
      dataOffset = pos + 8;
      dataSize = size;
      break;
    }
    pos += 8 + size;
  }
  if (sampleRate !== 16000 || channels !== 1 || bitsPerSample !== 16) {
    throw new Error(
      `${filePath}: must be 16 kHz mono 16-bit, got ${sampleRate} Hz / ` +
      `${channels} channel(s) / ${bitsPerSample}-bit`
    );
  }
  return new Int16Array(
    buf.buffer,
    buf.byteOffset + dataOffset,
    dataSize / 2,
  );
}

function main(modelPath, wavPath) {
  console.log("VoxRT wake-word — Node.js quickstart");
  console.log(`  model: ${modelPath}`);
  console.log(`  audio: ${wavPath}\n`);

  const engine = WakeWordEngine.fromPath(modelPath);
  engine.threshold = 0.9;
  engine.cooldownFrames = 100;

  const samples = readWavMono16(wavPath);
  console.log(`  duration: ${(samples.length / 16000).toFixed(2)} s\n`);

  const start = process.hrtime.bigint();
  let wakes = 0;
  for (let off = 0; off < samples.length; off += CHUNK_FRAMES) {
    const chunk = samples.subarray(off, off + CHUNK_FRAMES);
    for (const d of engine.pushPcmI16(chunk)) {
      wakes++;
      console.log(
        `  [wake] frame=${d.frameIndex} ` +
        `t=${d.timestampSec.toFixed(3)}s ` +
        `score=${d.score.toFixed(4)}`
      );
    }
  }
  const elapsedNs = Number(process.hrtime.bigint() - start);
  const elapsed = elapsedNs / 1e9;
  const audioSec = samples.length / 16000;

  console.log();
  console.log(
    `  processed ${audioSec.toFixed(2)} s of audio in ${elapsed.toFixed(3)} s wall-clock`
  );
  console.log(
    `  RTF = ${(elapsed / audioSec).toFixed(4)} (${wakes} wake event(s) emitted)`
  );
}

if (process.argv.length !== 4) {
  console.error("usage: node quickstart.js <model.vxrt> <audio.wav>");
  process.exit(1);
}
main(process.argv[2], process.argv[3]);

File: examples/python/quickstart/quickstart.py
"""voxrt-wake-word quickstart — load a .wav, run wake-word over it.

Demonstrates the smallest possible end-to-end pipeline:

  1. Construct WakeWordEngine from a .vxrt path.
  2. Read a 16 kHz mono WAV file in 512-sample chunks (32 ms each).
  3. Push each chunk through push_pcm_i16, print any detections.

Pure stdlib — uses `wave` to read the WAV so the example works on a
fresh Raspberry Pi without `numpy` / `soundfile` / `sounddevice`.

Usage:
    python3 quickstart.py voxrt_wake_word.vxrt my_recording.wav

The WAV must be 16 kHz, mono, 16-bit PCM. Convert with:
    ffmpeg -i input.flac -ar 16000 -ac 1 -sample_fmt s16 output.wav

For live-microphone streaming, install `sounddevice` (`pip install
sounddevice`) and feed `rec(samplerate=16000, channels=1, dtype='int16')`
chunks into `engine.push_pcm_i16` — same API.
"""

import struct
import sys
import wave
from time import perf_counter

from voxrt_wake_word import WakeWordEngine

CHUNK_FRAMES = 512  # 32 ms at 16 kHz


def main(model_path: str, wav_path: str) -> None:
    print(f"VoxRT wake-word — quickstart")
    print(f"  model: {model_path}")
    print(f"  audio: {wav_path}")
    print()

    engine = WakeWordEngine.from_path(model_path)
    engine.threshold = 0.9
    engine.cooldown_frames = 100

    with wave.open(wav_path, "rb") as wav:
        sr = wav.getframerate()
        channels = wav.getnchannels()
        sampwidth = wav.getsampwidth()
        if sr != 16000 or channels != 1 or sampwidth != 2:
            sys.exit(
                f"quickstart: WAV must be 16 kHz mono 16-bit — "
                f"got rate={sr} channels={channels} sampwidth={sampwidth}"
            )
        total_frames = wav.getnframes()
        print(f"  duration: {total_frames / sr:.2f} s")
        print()

        start = perf_counter()
        wakes = 0
        remaining = total_frames
        while remaining > 0:
            n = min(CHUNK_FRAMES, remaining)
            raw = wav.readframes(n)
            samples = struct.unpack(f"<{n}h", raw)
            for d in engine.push_pcm_i16(list(samples)):
                wakes += 1
                print(f"  [wake] frame={d.frame_index} "
                      f"t={d.timestamp_sec:.3f}s score={d.score:.4f}")
            remaining -= n
        elapsed = perf_counter() - start

    audio_sec = total_frames / 16000.0
    print()
    print(f"  processed {audio_sec:.2f} s of audio in {elapsed:.3f} s wall-clock")
    print(f"  RTF = {elapsed / audio_sec:.4f} ({wakes} wake event(s) emitted)")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        sys.exit("usage: quickstart.py <model.vxrt> <audio.wav>")
    main(sys.argv[1], sys.argv[2])
